sql/logictest: TestLogic failed

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4820263&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4820263&tab=artifacts#/) on master @ [9bff1de7115984957978416d829feb582eb2664e](https://github.com/cockroachdb/cockroach/commits/9bff1de7115984957978416d829feb582eb2664e):


```
=== RUN   TestLogic
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/7013a0eb507caf06d50eefd8e168b91f/logTestLogic4246381389
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestLogic
    logic.go:4241: -- test log scope end --
--- FAIL: TestLogic (5408.35s)
=== RUN   TestLogic/fakedist-metadata
    --- FAIL: TestLogic/fakedist-metadata (644.53s)
=== RUN   TestLogic/fakedist-metadata/cluster_locks
[06:09:45] rng seed: -3020448157605857873

[06:09:45] --- queries start here (file: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3569/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/logictest/logictest_test_/logictest_test.runfiles/cockroach/pkg/sql/logictest/testdata/logic_test/cluster_locks)
[06:09:45] CREATE TABLE t (k STRING PRIMARY KEY, v STRING);
[06:09:45] rewrote:
CREATE TABLE t (k STRING PRIMARY KEY, v STRING, FAMILY (v), FAMILY (k));


[06:09:45] GRANT ALL ON t TO testuser;
[06:09:45] INSERT INTO t VALUES ('a', 'val1'), ('b', 'val2'), ('c', 'val3'), ('l', 'val4'), ('m', 'val5'), ('p', 'val6'), ('s', 'val7'), ('t', 'val8'), ('z', 'val9');
[06:09:45] ALTER TABLE t SPLIT AT VALUES ('d'), ('r');
[06:09:45] 	-- OK;
[06:09:45] SELECT start_key, end_key, replicas, lease_holder FROM [SHOW RANGES FROM TABLE t];
[06:09:45] 	-- OK;
[06:09:45] BEGIN; UPDATE t SET v = concat(v, '_updated') WHERE k >= 'b' and k < 'z';
[06:09:45] BEGIN;
[06:09:45] select * from t for update;
[06:09:45] ROLLBACK;
[06:09:45] 	-- FAIL
    logic.go:3311: 
        
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3569/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/logictest/logictest_test_/logictest_test.runfiles/cockroach/pkg/sql/logictest/testdata/logic_test/cluster_locks:41: 
        expected success, but found
        (XXUUU) there is no transaction in progress
        exec_util.go:727: in init()
[06:09:45] 	-- OK;
[06:09:45] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3569/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/logictest/logictest_test_/logictest_test.runfiles/cockroach/pkg/sql/logictest/testdata/logic_test/cluster_locks with config fakedist-metadata: 8 tests, 1 failures

    logic.go:4009: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3569/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/logictest/logictest_test_/logictest_test.runfiles/cockroach/pkg/sql/logictest/testdata/logic_test/cluster_locks:41: error while processing
    logic.go:4009: pq: there is no transaction in progress
        --- FAIL: TestLogic/fakedist-metadata/cluster_locks (0.69s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-14908

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -3336,11 +3336,14 @@ func (t *logicTest) execStatement(stmt logicStatement) (bool, error) {
 		}
 		t.pendingStatements[stmt.statementName] = pending
 
+		startedChan := make(chan struct{})
 		go func() {
+			startedChan <- struct{}{}
 			res, err := t.db.Exec(execSQL)
 			pending.resultChan <- pendingExecResult{execSQL, res, err}
 		}()
 
+		<-startedChan
 		return true, nil
 	}
 

